### PR TITLE
Константинов Илья. Вариант 20. Технология OMP. Поразрядная сортировка для вещественных чисел (тип double) с четно-нечетным слиянием Бэтчера.

### PR DESCRIPTION
**Pull request by Ilya2643**
Программа сортирует массив чисел с плавающей точкой (типа double) методом поразрядной сортировки. Процесс состоит из нескольких этапов:

Валидация: Проверяется, что входной и выходной массивы имеют одинаковый размер. Если размеры не совпадают, программа должна сообщить об ошибке и прекратить работу.
Препроцессинг (Предварительная обработка): Входные данные копируются во внутреннюю структуру данных программы (класс задачи). Это делается для удобства обработки и предотвращения случайных изменений исходных данных.
3 Обработка (Сортировка): Здесь происходит основная часть работы — собственно, поразрядная сортировка чисел с плавающей точкой. Поразрядная сортировка работает по принципу последовательного распределения элементов по "корзинам" в зависимости от значения отдельных битов (или частей) числа. Так как это числа с плавающей точкой, необходимо учитывать особенности их представления в памяти (мантиссу и экспоненту), что делает реализацию сложнее, чем для целых чисел.
Постпроцессинг (Последующая обработка): Отсортированные данные из внутренней структуры программы записываются в выходной массив, который был передан в функцию сортировки. Таким образом, пользователь получает отсортированный массив в том же месте, где он предоставил исходный массив.
В целом, алгоритм обеспечивает сортировку массива чисел с плавающей точкой, при этом гарантируя целостность входных данных и корректность выходного результата. Сложность реализации поразрядной сортировки для чисел с плавающей запятой обусловлена необходимостью обработки их внутреннего представления (мантисса и порядок). 
Добавлена реализация распараллеливания за счет openMP в циклах с переводом в двоичное представление и обратно, а также параллельный подсчет счетчика байт для числа.

### CODE REVIEW

**Review by odincovm — CHANGES_REQUESTED** [review]
add a test for an empty array and a test for an array with already sorted data to check the implementation of the code.

**Review by ermilovad — CHANGES_REQUESTED** [review]
add a test for an array with back-sorted data and a test for an array with duplicate values to check the implementation of the code

**Review by odincovm — APPROVED** [review]

**Review by ermilovad — APPROVED** [review]

File: tasks/omp/Konstantinov_I_Sort_Batcher/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  konstantinov_i_sort_batcher_omp::RadixSortBatcherOmp test_task_omp(task_data_omp);
+  ASSERT_EQ(test_task_omp.ValidationImpl(), true);
+  test_task_omp.PreProcessingImpl();
+  test_task_omp.RunImpl();
+  test_task_omp.PostProcessingImpl();
+  EXPECT_EQ(exp_out, out);
+}
+
+TEST(Konstantinov_I_Sort_Batcher_omp, test_random_10000_values) {
+  constexpr size_t kCount = 10000;
```
Comment (allnes) [inline tasks/omp/Konstantinov_I_Sort_Batcher/func_tests/main.cpp:171]: please check 11887 elements

**Comment by allnes** [comment]
@Ilya2643 please go through the review from another student.

**Review by aobolensk — APPROVED** [review]

File: tasks/omp/Konstantinov_I_Sort_Batcher/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <random>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "omp/Konstantinov_I_Sort_Batcher/include/ops_omp.hpp"
+
+TEST(Konstantinov_I_Sort_Batcher_omp, test_empty_array) {
```
Comment (Semyon1104) [inline tasks/omp/Konstantinov_I_Sort_Batcher/func_tests/main.cpp:13]: add a test that checks how sorting works on an array with alternating positive and negative numbers

Comment (Ilya2643) [reply tasks/omp/Konstantinov_I_Sort_Batcher/func_tests/main.cpp:13]: added as requested

**Review by Semyon1104 — APPROVED** [review]

**Review by allnes — APPROVED** [review]
